xcode/Subconscious/Shared/Library/Prose.swift

```diff
@@ -0,0 +1,66 @@
+//
+//  Prose.swift
+//  Subconscious (iOS)
+//
+//  Created by Gordon Brander on 3/2/23.
+//
+
+import Foundation
+
+extension Character {
+    private static let sentenceEndingPunctuation: Set<Character> = Set(
+        [
+            ".",
+            "!",
+            "?",
+            "\n"
+        ]
+    )
+    
+    var isSentenceEndingPunctuation: Bool {
+        Self.sentenceEndingPunctuation.contains(self)
+    }
+}
+
+extension String {
+    /// Truncate string to max length, appending ellipsis if truncated.
+    func truncate(
+        maxLength: Int = 256,
+        ellipsis: String = "…",
+        fallback: String = ""
+    ) -> String {
+        guard self.count > maxLength else {
+            return self
+        }
+        let truncated = self.prefix(
+            maxLength
+        ).trimmingCharacters(
+            in: .whitespacesAndNewlines
+        )
+        return !truncated.isEmpty ? "\(truncated)\(ellipsis)" : ""
```

Comment:
If I pass `maxLength` = `256` does this result in a `257` character string due to the ellipsis?

Reply:
Correct. Since ellipsis can be any string, truncating to x chars INCLUDING ellipsis requires a lot of additional logic to make sure you stay under budget.

I could change the label to `maxLengthNotIncludingEllipsis` or `maxTextLength`?

Reply:
Renaming to `maxLengthNotIncludingEllipsis` works for me me but I would've though that `self.prefix(maxLength - ellipsis.count)` would do the job?

Reply:
... which raises a question for me of what this function does when passed `maxLength` of `0` or `-1` etc.

Reply:
Actually you're right and its just late for me. I'm too fried for this rn. Fixed and updated tests.